torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -411,44 +411,84 @@ def _step_microbatches(
         fwd_sends_to_wait: List[dist.Work] = []
         bwd_sends_to_wait: List[dist.Work] = []
 
+        def is_forward_step(i):
+            assert i >= 0, i
+            return i < self._n_microbatches
+
+        def is_backward_step(i):
+            assert i < total_steps, i
+            return i >= warmup_steps and self._has_backward
+
+        def is_1f1b_step(i):
+            return is_forward_step(i) and is_backward_step(i)
```

Comment:
nit: I think a "step" in this schedule class means either a microbatch forward or a microbatch backward.
Thus, `is_forward_step(i) and is_backward_step(i)` is not something very straightforward to understand, like "how can a step be a forward and a backward at the same time?"
Although, if looking deeper, one would understand the logic here.

Reply:
do you think it would be clearer if i changed the name to `step_has_forward`?

Reply:
fixing on later PR on the stack. trying to land now